src/Reflection/Apply.agda

```diff
@@ -0,0 +1,200 @@
+      with does (v₁ ≟ᵥ v₂)
+  ...    | true = subst 0 x t
+  ...    | false = failed-app b a
+  -- catch all
+  app a t = failed-app t a
+
+  app-many [] t = return t
+  app-many (a ∷ as) t = do
+    ta ← app a t
+    app-many as ta
+
+  subst i x (var j args) with compare i j
+  ...                       | less m k = do
+                              args ← subst-args i x args
+                              -- decrement j by one since λ was eliminated
+                              return $ var (m + k) args -- j ≡ suc (m + k)
+  ...                       | greater _ _ = do
+                              args ← subst-args i x args
+                              -- j refers to variable named inside eliminated λ
+                              return $ var j args
+  subst ⦃ fuel (suc n) ⦄
+        i x (var j args)    | equal _ = do
+                              args ← subst-args ⦃ fuel (suc n) ⦄ i x args
+                              app-many ⦃ fuel n ⦄ args (shift-index i 0 x)
+  subst ⦃ fuel zero ⦄
+        i x (var j [])      | equal _ = return x
+  subst ⦃ fuel zero ⦄
+        _ _ (var j (_ ∷ _)) | equal _ = err (strErr "evaluation limit reached" ∷ [])
+  subst i x (con c args) = do
+    args ← subst-args i x args
+    return $ con c args
+  subst i x (def f args) = do
+    args ← subst-args i x args
+    return $ def f args
+  subst i x (lam v (abs s t)) = do
+    t ← subst (suc i) x t
+    return $ lam v (abs s t)
+  subst i x (meta m args) = do
+    args ← subst-args i x args
+    return $ meta m args
+  subst i x (sort s) = return $ sort s
+  subst i x (lit l) = return $ lit l
+  subst i x unknown = return unknown
+  subst i x (pat-lam cs args) = do
+    cs ← subst-clauses i x cs
+    args ← subst-args i x args
+    return $ pat-lam cs args
+  subst i x (Π[ s ∶ arg v A ] t) = do
+    A ← subst i x A
+    t ← subst (suc i) x t
+    return $ Π[ s ∶ arg v A ] t
+
+  subst-args i x [] = return []
+  subst-args i x (arg v a ∷ as) = do
+    a ← subst i x a
+    as ← subst-args i x as
+    return $ arg v a ∷ as
+
+  subst-clauses i x [] = return []
+  subst-clauses i x (clause ps t ∷ cs) = do
+    t ← subst (i + pattern-args-size ps) x t
+    cs ← subst-clauses i x cs
+    return $ clause ps t ∷ cs
+  subst-clauses i x (absurd-clause ps ∷ cs) = do
+    cs ← subst-clauses i x cs
+    return $ absurd-clause ps ∷ cs
+
+{- Apply an argument to a term. Fails if the recusion limit is reached or an
+   attempt is made to apply an argument to non-function-like term.
+-}
+apply : Term → Arg Term → Result Term
+apply f x = app ⦃ fuel limit ⦄ x f
+
+open import Relation.Binary.PropositionalEquality
+
+open import Relation.Unary
+
+{- Retrieve any trailing arguments in a term -}
+term-args : Term → Maybe (Args Term)
+term-args (var _ args) = just args
+term-args (con _ args) = just args
+term-args (def _ args) = just args
+term-args (pat-lam _ args) = just args
+term-args (meta _ args) = just args
+term-args other = nothing
```

Comment:
Any chance we could align the `=`?

Reply:
My space-bar says it is up to the task, yes.